README.md
<div align="center">
<a href="https://www.cemac.leeds.ac.uk/">
  <img src="https://github.com/cemac/cemac_generic/blob/master/Images/cemac.png"></a>
  <br>
</div>

# CEMAC FORK OF MIKE CROUCHERS JUPYERHUB ON AZURE
 
 :construction: MAJOR OVERHAUL OF DOCUMENTATION IN PROGRESS

[![GitHub release](https://img.shields.io/badge/release-v.2.0-blue.svg)](
https://github.com/cemac/jupyterhub_on_azure/releases/tag/2.0)

This repository contains scripts for automatic creation of customised JupyterHub instances running on Azure cloud designed for multi-user classes.

```diff
- NB: This is designed to work following Leeds Uni Practice
- Central IT control the Azure account and must set up VM (Jan 2019)
```

# Version 2.0 #

Following on from Mike Croucher's jupyterhub_on_azure setup, Version builds on documentation and functionality.

## Features ##

* Set up JupyterHub on Azure Ubuntu LTS machine
* Server settings
* Anaconda Installation
* Jupyter hub configuration
* Certificate generation
* Legacy VM creation scripts

## Requirements

* Azure VM, IP address and Admin ID to access (via IT following [create_vm](create_vm))
* UBUNTU 18.04 LTS (*recommended*)

## Installation/ Usage (recommended method)

* ssh to Azure machine `ssh -Y <adim_ID>@<VM_IP>`
* `git clone https://github.com/cemac/jupyterhub_on_azure.git`
* Follow [ubuntu_setup](./ubuntu_setup) to set up server and users
* Follow [python_setup](./python_setup) to set up python environment
* Follow [jupyterhub_setup](./jupyterhub_setup) to set up Jupyter Hub

*NB previously this set up was done using mike's create_vm scripts, however this
is now managed by Leeds central IT - the scripts are preserved for reference in [create_vm](create_vm)*

## Tips ##


## COLLABORATING

Collaboration is very welcome and encouraged.
* A full guide on submitting issue and bug are given in our [contribution guidelines](https://github.com/cemac/jupyterhub_on_azure/blob/master/CONTRIBUTING.md)
* Please use our issue templates for feature requests and bug fixes.

<hr>

## FUTURE ##

This project will move to a containerised set up in the future in a new JupyterHubLeeds
project for academic year 2019/2020.

<hr>

## ACKNOWLEGEMENTS

Thanks to Mike Croucher for Setting this up.

## LICENSE

This project is currently licensed under the [MIT license](https://choosealicense.com/licenses/mit/).


history.txt
cat /proc/cpuinfo
cat /proc/meminfo
df -h
sudo apt install emacs25
emacs -nw .bashrc
emacs .emacs.d/init.el
sudo apt-get update
sudo apt-get upgrade
ufw status
sudo ufw status
sudo ufw allow OpenSSH
sudo ufw enable
sudo emacs -nw /etc/default/ufw
sudo ufw allow 22
sudo ufw enable
sudo ufw allow ssh
sudo ufw enable
sudo ufw status
AZ_REPO=$(lsb_release -cs)
echo $AZ_REPO
echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $AZ_REPO main" |     sudo tee /etc/apt/sources.list.d/azure-cli.list
echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $AZ_REPO main" |     sudo tee /etc/apt/sources.list.d/azure-cli.list
curl -L https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
sudo apt-get install apt-transport-https
sudo apt-get update && sudo apt-get install azure-cli
sudo sed -i s/DIR_MODE=0755/DIR_MODE=0750/g /etc/adduser.conf
# Add the aliases requested from the academics to .bashrc
sudo apt-get -y install apg
wget https://repo.anaconda.com/archive/Anaconda3-latest-Linux-x86_64.sh
sudo chmod ugo+w /opt/anaconda3
sudo emacs /etc/profile
sudo chmod -R g+w anaconda3/
sudo chown -R root:jupyteradmin anaconda3
ps aux | grep configurable-http-proxy
sudo systemctl daemon-reload
sudo systemctl restart jupyterhub
groups
usermod -a -G jupyterhub geosciences_instructor
usermod -a -G shadow jupyterhub
sudo -H -u geosciences_instructor /opt/anaconda3/bin/jupyter nbextension enable --user create_assignment/main
sudo systemctl daemon-reload
sudo cp ./jupyterhub.service /etc/systemd/system/jupyterhub.service
sudo systemctl daemon-reload
journalctl -u jupyterhub.service
sudo groupadd -g 1160 jupyterhub
sudo usermod -aG jupyterhub geosciences_instructor
sudo rsnapshot configtest
sudo rsnapshot -t hourly
sudo emacs -nw /etc/rsnapshot.conf
sudo rsnapshot -t alpha
sudo chown -R jupyteradmin:jupyteradmin /storage/backup/
sudo rsnapshot -t alpha
sudo rsnapshot help
sudo chgrp geosciences_instructor /storage/earth_data/
sudo chmod g+rwx /storage/earth_data/
sudo emacs -nw /etc/rsnapshot.conf
chmod g+rw earth_data/
history > history.txt


install.sh
#!/bin/bash
#Install Azure CLI
AZ_REPO=$(lsb_release -cs)
echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $AZ_REPO main" | \
    sudo tee /etc/apt/sources.list.d/azure-cli.list
curl -L https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
sudo apt-get install apt-transport-https
sudo apt-get update && sudo apt-get install azure-cli

#By default, users can read the files in each other's home directory.
#Change this so only sudo users have this ability
sudo sed -i s/DIR_MODE=0755/DIR_MODE=0750/g /etc/adduser.conf

# Add the aliases requested from the academics to .bashrc
sudo cat << EOF >> /etc/skel/.bashrc
alias rm='rm -i'
alias mv='mv -i'
alias cp='cp -i'
EOF

#create users
number_of_users=4
password_file=~/users.txt

#Install apg for password generation
sudo apt-get -y install apg

#On Azure, the following creates errors like
#sent invalidate(group) request, exiting
#sent invalidate(passwd) request, exiting
#This shouldn't be a problem
 touch $password_file
 for i in `seq 1 $number_of_users`;
  do
  username=training_user$i
  sudo adduser --disabled-password --gecos "" $username
  userpassword=`apg -n 1`
  echo $username:$userpassword | sudo chpasswd
  echo "UserID:" $username "has been created with the following password " $userpassword >> $password_file
 done

 #Install the instructor username
 sudo adduser --disabled-password --gecos "" instructor
 userpassword=`apg -n 1`
 echo instructor:$userpassword | sudo chpasswd
 echo "UserID:" instructor "has been created with the following password " $userpassword >> $password_file

#Install pre-reqs for jupyterhub
sudo apt-get -y install npm
sudo apt-get -y install python3-pip
sudo apt-get -y install nodejs
#Update node
sudo npm cache clean -f
sudo npm install -g n
sudo n stable
#install configurable-http-proxy
sudo npm install -g configurable-http-proxy

#Install Jupyterhub
python3 -m pip install jupyterhub
sudo python3 -m pip install sudospawner

#Install the notebook
sudo python3 -m pip install notebook

#Installing by pip didn't add the jupyterhub binary folder to the path so let's do that now
export PATH=$PATH:~/.local/bin

#Generate default config
jupyterhub --generate-config

#Install some packages
sudo python3 -m pip install numpy scipy matplotlib
#Packages for earth and environment
#cartopy
sudo apt-get -y install libproj-dev proj-data proj-bin
sudo apt-get -y install libgeos-dev
sudo python3 -m pip install cython
sudo python3 -m pip install cartopy
#netcdf
sudo apt-get -y install libnetcdf-dev netcdf-bin
sudo python3 -m pip install netCDF4

# Install nbgrader
sudo python3 -m pip install nbgrader
jupyter nbextension install --system --py nbgrader --overwrite
jupyter nbextension enable --system --py nbgrader
jupyter serverextension enable --system --py nbgrader
#Disable create assignment for all users and then enable it for instructor only
#Following the prodedure at https://nbgrader.readthedocs.io/en/latest/user_guide/installation.html retrieved 11/12/2018
jupyter nbextension disable --sys-prefix create_assignment/main
sudo -H -u instructor jupyter nbextension enable --user create_assignment/main

# Move certificate files
secretsname=$(sudo find /var/lib/waagent/ -name "*.prv" | cut -c -57)
sudo mkdir -p /etc/jupyter/ssl
sudo cp $secretsname.crt /etc/jupyter/ssl/mycert.cert
sudo cp $secretsname.prv /etc/jupyter/ssl/mycert.prv

#Make certificate files readable by the user under which we will run the jupyterhub service
sudo chgrp azureuser /etc/jupyter/ssl/mycert.prv
sudo chmod g+r /etc/jupyter/ssl/mycert.cert
sudo chmod g+r /etc/jupyter/ssl/mycert.prv

#Configure certificate in Jupyterhub
cat << EOF >> ./jupyterhub_config.py
c.JupyterHub.ssl_key = '/etc/jupyter/ssl/mycert.prv'
c.JupyterHub.ssl_cert = '/etc/jupyter/ssl/mycert.cert'
EOF

#Run on port 443 so that it uses https
cat << EOF >> ./jupyterhub_config.py
c.JupyterHub.port = 443
EOF

#Redirect http:// to https://
cat << EOF >> ./jupyterhub_config.py
c.ConfigurableHTTPProxy.command = ['configurable-http-proxy', '--redirect-port', '80']
EOF

#Add jupyterhub admin users
cat << EOF >> ./jupyterhub_config.py
c.Authenticator.admin_users = {'azureuser', 'instructor'}
EOF

#copy the Jupyterhub config file
sudo mkdir -p /etc/jupyterhub
sudo cp ./jupyterhub_config.py /etc/jupyterhub/jupyterhub_config.py
sudo chown -R azureuser:azureuser /etc/jupyterhub/

#Set up sudospawner
#Following docs at https://github.com/jupyterhub/jupyterhub/wiki/Using-sudo-to-run-JupyterHub-without-root-privileges retrieved 19th September 2018
echo "Cmnd_Alias JUPYTER_CMD=/usr/local/bin/sudospawner" | sudo tee -a /etc/sudoers
echo "%jupyterhub ALL=(azureuser) /usr/bin/sudo" | sudo tee -a /etc/sudoers
echo "azureuser ALL=(%jupyterhub) NOPASSWD:JUPYTER_CMD" | sudo tee -a /etc/sudoers

#Set up jupyterhub as a service
sudo cp ./jupyterhub.service /etc/systemd/system/jupyterhub.service

#Make our user part of the shadow group so that PAM authentication works
sudo usermod -a -G shadow azureuser

#Do this next line or we'll not be able to connect to port 443
#Details at https://github.com/jupyterhub/jupyterhub/issues/774
sudo setcap 'cap_net_bind_service=+ep' `which nodejs`
sudo setcap 'cap_net_bind_service=+ep' `which node`

#Enable the jupyterhub service so it starts at boot
sudo systemctl enable jupyterhub
#start the service now
sudo systemctl start jupyterhubureuser /etc/jupyter/ssl/mycert.cert
sudo chgrp azureuser /etc/jupyter/ssl/mycert.prv
sudo chmod g+r /etc/jupyter/ssl/mycert.cert
sudo chmod g+r /etc/jupyter/ssl/mycert.prv

#Configure certificate in Jupyterhub
cat << EOF >> ./jupyterhub_config.py
c.JupyterHub.ssl_key = '/etc/jupyter/ssl/mycert.prv'
c.JupyterHub.ssl_cert = '/etc/jupyter/ssl/mycert.cert'
EOF

#Run on port 443 so that it uses https
cat << EOF >> ./jupyterhub_config.py
c.JupyterHub.port = 443
EOF

#Redirect http:// to https://
cat << EOF >> ./jupyterhub_config.py
c.ConfigurableHTTPProxy.command = ['configurable-http-proxy', '--redirect-port', '80']
EOF

#Add jupyterhub admin users
cat << EOF >> ./jupyterhub_config.py
c.Authenticator.admin_users = {'azureuser', 'instructor'}
EOF

#copy the Jupyterhub config file
sudo mkdir -p /etc/jupyterhub
sudo cp ./jupyterhub_config.py /etc/jupyterhub/jupyterhub_config.py
sudo chown -R azureuser:azureuser /etc/jupyterhub/

#Set up sudospawner
#Following docs at https://github.com/jupyterhub/jupyterhub/wiki/Using-sudo-to-run-JupyterHub-without-root-privileges retrieved 19th September 2018
echo "Cmnd_Alias JUPYTER_CMD=/usr/local/bin/sudospawner" | sudo tee -a /etc/sudoers
echo "%jupyterhub ALL=(azureuser) /usr/bin/sudo" | sudo tee -a /etc/sudoers
echo "azureuser ALL=(%jupyterhub) NOPASSWD:JUPYTER_CMD" | sudo tee -a /etc/sudoers

#Set up jupyterhub as a service
sudo cp ./jupyterhub.service /etc/systemd/system/jupyterhub.service

#Make our user part of the shadow group so that PAM authentication works
sudo usermod -a -G shadow azureuser

#Do this next line or we'll not be able to connect to port 443
#Details at https://github.com/jupyterhub/jupyterhub/issues/774
sudo setcap 'cap_net_bind_service=+ep' `which nodejs`
sudo setcap 'cap_net_bind_service=+ep' `which node`

#Enable the jupyterhub service so it starts at boot
sudo systemctl enable jupyterhub
#start the service now
sudo systemctl start jupyterhub

#Connect the data drive
#sudo mkdir /datadrive
#The datadrive is probably going to be connected /dev/sdc1 but I should come up with some way of checking properly
#sudo mount /dev/sdc1 /datadrive

#Connect the backup drive
#sudo mkdir /backup
#The backup drive is probably going to be connected /dev/sdc1 but I should come up with some way of checking properly
#sudo mount /dev/sdd1 /backup

#Install rsnapshot to do the backups
sudo apt-get -y install rsnapshot
#Snapshot_root location
sudo sed -i s,/var/cache/rsnapshot/,/backup/,g /etc/rsnapshot.conf
#We don't want to back up /etc and /usr/local so comment these lines out
sudo sed -i 's,backup\t/etc,#backup\t/etc,g' /etc/rsnapshot.conf
sudo sed -i 's,backup\t/usr/local,#backup /usr/local,g' /etc/rsnapshot.conf
#Activate the cron job by uncommenting the relevant lines
sed -i '/alpha/s/^#//g' /etc/cron.d/rsnapshot
sed -i '/beta/s/^#//g' /etc/cron.d/rsnapshot
sed -i '/gamma/s/^#//g' /etc/cron.d/rsnapshot
